sn_client/src/wallet.rs

```diff
@@ -104,86 +103,103 @@ impl WalletClient {
-            self.store_cost().as_nano() * (2.0f64.powf((num_of_addrs / 100 + 1) as f64)) as u64,
-        );
+        while let Some(res) = tasks.join_next().await {
+            let (content_addr, amounts_to_pay) = match res {
+                Ok(c) => c,
+                Err(e) => return Err(Error::CouldNotGetStoreCost(e.to_string())),
+            };
+            // let amounts_to_pay = amounts_to_pay;
+            payment_map.insert(content_addr, amounts_to_pay?);
+        }
 
-        info!("Storage cost per record: {}", storage_cost);
+        let (payments, cost) = self.pay_for_records(payment_map, verify_store).await?;
 
-        let amount_to_pay = number_of_records_to_pay * storage_cost.as_nano();
+        if let Some(cost) = total_cost.checked_add(cost) {
+            total_cost = cost;
+        }
 
-        trace!("Making payment for {num_of_addrs} addresses of {amount_to_pay:?} nano tokens.");
+        Ok((payments, total_cost))
+    }
 
-        let transfer = self
+    /// Send tokens to nodes closest to the data we want to make storage payment for.
+    ///
+    /// Returns DbcIds created for the payment and the total amount paid
+    ///
+    /// This can optionally verify the store has been successful (this will attempt to GET the dbc from the network)
+    pub async fn pay_for_records(
+        &mut self,
+        all_data_payments: BTreeMap<NetworkAddress, Vec<(PublicAddress, Token)>>,
+        verify_store: bool,
+    ) -> Result<(BTreeMap<NetworkAddress, TransferOutputs>, Token)> {
+        let now = Instant::now();
+
+        let mut total_cost = Token::zero();
+        for (_data, costs) in all_data_payments.iter() {
+            for (_target, cost) in costs {
+                if let Some(new_cost) = total_cost.checked_add(*cost) {
+                    total_cost = new_cost;
+                } else {
+                    return Err(Error::TotalPriceTooHigh);
+                }
+            }
+        }
+        let transfer_map = self
             .wallet
-            .local_send_storage_payment(Token::from_nano(amount_to_pay), root_hash, None)
+            .local_send_storage_payment(all_data_payments, None)
             .await?;
 
         // send to network
-        trace!("Sending storage payment transfer to the network: {transfer:#?}");
-        if let Err(error) = self.client.send(transfer.clone(), verify_store).await {
-            warn!("The storage payment transfer was not successfully registered in the network: {error:?}. It will be retried later.");
-            self.unconfirmed_txs.push(transfer);
-            return Err(error);
+        trace!("Sending storage payment transfer to the network");
+
+        for transfer in transfer_map.values() {
+            // first push all to the network, then try again and verify.
+            // otherwise all pushes are waiting on all prior verifications
+            self.client.send(transfer.clone(), false).await?;
+
         }
```

Comment:
Maybe for the follow up PR? I think we can parallelize the `sends` here?

Reply:
hmm, things get complicated here, as they may actually be chained. I thin I'm going to remove this and we have to look at what makes sense to do in parallel here as optimisations down the line.